packages/core/src/Cardano/types/Strings/util.ts

```diff
@@ -0,0 +1,59 @@
+import { CustomError } from 'ts-custom-error';
+import { Decoded, bech32 } from 'bech32';
+
+// Source: https://github.com/Microsoft/Typescript/issues/202#issuecomment-811246768
+export declare class OpaqueString<T extends string> extends String {
+  /** This helps typescript distinguish different opaque string types. */
+  protected readonly __opaqueString: T;
+  /**
+   * This object is already a string, but calling this makes method
+   * makes typescript recognize it as such.
+   */
+  toString(): string;
+}
+
+export class InvalidStringError extends CustomError {
+  constructor(expectation: string) {
+    super(`Invalid string: "${expectation}"`);
+  }
+}
+
+// TODO: test this
+/**
+ * @param {string} target bech32 string to decode
+ * @param {string} prefix expected prefix
+ * @param {string} expectedDecodedLength number of expected words, >0
+ * @throws {InvalidStringError}
+ */
+export const assertIsBech32WithPrefix = (target: string, prefix: string, expectedDecodedLength?: number): void => {
+  let decoded: Decoded;
+  try {
+    decoded = bech32.decode(target);
+  } catch {
+    throw new InvalidStringError(`expected bech32-encoded string with '${prefix}' prefix`);
+  }
+  if (decoded.prefix !== prefix) {
+    throw new InvalidStringError(`expected bech32 prefix '${prefix}', got '${decoded.prefix}''`);
```

Comment:
This isn't covered by a test currently, and it's not clear to me why the check is required